[pytorch][quant][oss] Support 2-bit embedding_bag op "embedding_bag_2bit_rowwise_offsets"

ShijunK opened the pull request:
Summary: Support 2-bit embedding_bag op "embedding_bag_2bit_rowwise_offsets"

Test Plan:
```
buck test caffe2/test/:quantization -- test_embedding_bag_
Downloaded 0/2 artifacts, 0.00 bytes, 100.0% cache miss (for updated rules)
Building: finished in 6.7 sec (100%) 12614/12614 jobs, 2/12614 updated
  Total time: 7.1 sec
More details at https://www.internalfb.com/intern/buck/build/70dd2b34-5804-467f-937b-d9fd97b1cf7e
Tpx test run coordinator for Facebook. See https://fburl.com/tpx for details.
Running with tpx session id: 6962973e-2d54-4738-b1fb-8d309d2318ef
Trace available for this run at /tmp/tpx-20210820-075327.194248/trace.log
Started reporting to test run: https://www.internalfb.com/intern/testinfra/testrun/6192449499382619
    ✓ ListingSuccess: caffe2/test:quantization - main (10.681)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_2d_indices (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (9.902)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_2bit (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (9.949)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_byte (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (10.263)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_4bit (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (10.291)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_byte_unpack (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (10.527)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_4bit_unpack (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (10.585)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_2bit_unpack (quantization.core.test_quantized_op.TestQuantizedEmbeddingOps) (10.746)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_padding_idx_error (quantization.jit.test_quantize_jit.TestQuantizeDynamicJitOps) (15.127)
    ✓ Pass: caffe2/test:quantization - test_embedding_bag_api (quantization.core.test_quantized_module.TestStaticQuantizedModule) (21.161)
Summary
  Pass: 9
  ListingSuccess: 1
If you need help understanding your runs, please follow the wiki: https://fburl.com/posting_in_tpx_users
Finished test run: https://www.internalfb.com/intern/testinfra/testrun/6192449499382619
```

Differential Revision: D30454994

supriyar commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -3479,6 +3482,33 @@ def test_embedding_bag_4bit(self, num_embeddings,
+           include_last_offset=st.booleans(),
+           fallback_to_no_sparse=st.booleans(),
+           sparsity=st.sampled_from([0.0, 0.5, 0.7]))
+    def test_embedding_bag_2bit(self, num_embeddings,
```
Are we also testing FP16 inputs in this test?

ShijunK replied:
FP16 makes sense for prepack/unpack? which is covered in 

```
 """ Tests the correctness of the embedding_bag_8bit pack/unpack op against C2 """
    @given(num_embeddings=st.integers(10, 100),
           embedding_dim=st.integers(5, 50).filter(lambda x: x % 4 == 0),
           num_batches=st.integers(1, 5),
           data_type=st.sampled_from([np.float32, np.float16]),)
    def test_embedding_bag_byte_unpack(self, num_embeddings, embedding_dim, num_batches, data_type):
        pack_fn = torch.ops.quantized.embedding_bag_byte_prepack
        unpack_fn = torch.ops.quantized.embedding_bag_byte_unpack

        self._test_embedding_bag_unpack_fn(
            pack_fn, unpack_fn, num_embeddings, embedding_dim, 8, False, num_batches, data_type=data_type)
```

supriyar commented:
It looks like there is an ASAN failure in one of the tests, @ShijunK can you take a look at that?

jingsh commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -141,9 +141,10 @@ at::Tensor& embedding_lookup_fallback_impl(
+at::Tensor& embedding_bag_nbit_impl(
     at::Tensor& output,
     const at::Tensor& weight,
+    int bit_width,
```
```suggestion
    const int bit_width,
```

make this arg `const`

jingsh commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -174,8 +175,9 @@ at::Tensor& embedding_bag_4bit_impl(
 
   const auto weight_sizes = weight.sizes();
   const int64_t weight_size = weight_sizes[1];
+  int NUM_ELEM_PER_BYTE = 8 / bit_width;
```
~~should we add assert that `bit_width` must be > 0 and <= 8, and can only be multiples of 2? So the `bit_width` can only be 2, 4, or 8~~

nvm, noticed the assert is done below

jingsh commented on aten/src/ATen/native/quantized/cpu/qembeddingbag.cpp:
```diff
@@ -519,16 +539,21 @@ at::Tensor& embedding_bag_byte_helper(
+at::Tensor& _embedding_bag_nbit_helper(
     at::Tensor& output,
     const at::Tensor& weight,
+    int bit_width,
```
```suggestion
    const int bit_width,
```

ShijunK committed:
[pytorch][quant][oss] Support 2-bit embedding_bag op "embedding_bag_2bit_rowwise_offsets" (#63658)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/63658

Support 2-bit embedding_bag op "embedding_bag_2bit_rowwise_offsets"

Reviewed By: jingsh, supriyar

Differential Revision: D30454994

fbshipit-source-id: 457ee1ac1143a7b05ed317848fa698fc186cde71
